crates/mdns/src/lib.rs

```diff
@@ -1,172 +1,435 @@
+    pub fn clear_buffer(&mut self) {
+        self.buffer.zeroize();
+    }
+
+    /// Clean our cache by removing the out of live records.
+    pub fn prune_cache(&mut self) {
+        // Get the entry of the dead records to remove them safely afterward
+        // let mut dead_entry_list: Vec<String> = Vec::with_capacity(self.map_record.len());
+        for (_, records) in self.map_record.iter_mut() {
+            let _: Vec<Record> = records.drain_filter(|r| r.ttl == 0).collect();
+        }
+    }
+
+    /// Update the `time to live` of every cached record.
+    pub fn update_ttl(&mut self) {
+        let own_urls = self.own_urls();
+        for (_netid, records) in self.map_record.iter_mut() {
+            for record in records {
+                if !own_urls.contains(&record.url) {
+                    if record.ttl > 0 {
+                        record.ttl -= 1;
+                    }
+                }
+            }
+        }
+    }
+
+    /// mDNS Querier
+    /// One-Shot Multicast DNS Queries
+    /// Not used in our actual mDNS implementation use case.
+    pub fn query(&mut self) -> MulticastDnsResult<()> {
+        if let Some(query_message) = self.build_query_message() {
+            self.broadcast_message(&query_message)?;
+        }
+
+        Ok(())
+    }
+
+    /// A mDNS Responder that listen to the network in order to defend its name and respond to
+    /// queries
+    /// Not used in our actual mDNS implementation use case.
+    fn responder(&mut self) -> MulticastDnsResult<()> {
+        // Process all elements of the UDP socket stack
+        loop {
+            match self.recv() {
+                Ok(Some((packet, sender_addr))) => {
+                    let dmesg = DnsMessage::from_raw(&packet)?;
+
+                    // Here we update our cache with the responses gathered from the network
+                    if dmesg.nb_answers > 0 {
+                        if let Some(new_map_record) = MapRecord::from_dns_message(&dmesg) {
+                            self.update_cache(&new_map_record);
+                        }
+                    }
+                    // We send response only for record we have authority on.
+                    // We send the response directly to the sender instead of broadcasting it to
+                    // avoid any unnecessary burden on the network.
+                    if dmesg.nb_questions > 0 {
+                        let question_list: Vec<&str> = dmesg
+                            .questions
+                            .iter()
+                            .map(|q| q.domain_name.as_str())
+                            .collect();
+                        if let Some(response) =
+                            self.own_map_record.to_dns_response_message(&question_list)
+                        {
+                            self.broadcast_message(&response)?;
+                            self.send_socket.send_to(&response.to_raw()?, sender_addr)?;
```

Comment:
The comment above makes it seem like we don't broadcast, but then there's a broadcast here.

For the record, I'm personally fine broadcasting on the LAN since we tend to have more throughput locally, do you happen to know what the mdns spec recommends?